src/unix/timer.c

```diff
@@ -0,0 +1,655 @@
+                ut->cid, tinit, absolute, interval);
+    timer t = register_timer(ut->cid, tinit, absolute, interval,
+                             closure(unix_timer_heap, timerfd_timer_expire, ut));
+    if (t == INVALID_ADDRESS)
+        return -ENOMEM;
+
+    ut->t = t;
+    return 0;
+}
+
+sysreturn timerfd_gettime(int fd, struct itimerspec *curr_value)
+{
+    unix_timer ut = resolve_fd(current->p, fd); /* macro, may return EBADF */
+    if (ut->f.type != FDESC_TYPE_TIMERFD)
+        return -EINVAL;
+
+    if (!curr_value)
+        return -EFAULT;
+
+    itimerspec_from_timer(ut, curr_value);
+    return 0;
+}
+
+closure_function(5, 1, sysreturn, timerfd_read_bh,
+                 unix_timer, ut, void *, dest, u64, length, thread, t, io_completion, completion,
+                 u64, flags)
+{
+    unix_timer ut = bound(ut);
+    thread t = bound(t);
+    boolean blocked = (flags & BLOCKQ_ACTION_BLOCKED) != 0;
+    sysreturn rv = sizeof(u64);
+
+    timer_debug("fd %d, dest %p, length %ld, tid %d, flags 0x%lx\n",
+                ut->info.timerfd.fd, bound(dest), bound(length), t->tid, flags);
+
+    if (bound(length) < sizeof(u64)) {
+        assert(!blocked);
+        rv = -EINVAL;
+        goto out;
+    }
+
+    if (flags & BLOCKQ_ACTION_NULLIFY) {
+        assert(blocked);
+        rv = -EINTR;
+        goto out;
+    }
+
+    if (ut->info.timerfd.canceled) {
+        rv = -ECANCELED;
+        goto out;
+    }
+
+    u64 overruns = ut->overruns;
+    if (overruns == 0) {
+        if (!blocked && (ut->f.flags & TFD_NONBLOCK)) {
+            rv = -EAGAIN;
+            goto out;
+        }
+        timer_debug("   -> block\n");
+        return BLOCKQ_BLOCK_REQUIRED;
+    }
+    *(u64*)bound(dest) = overruns;
+    ut->overruns = 0;
+  out:
+    timer_debug("   -> returning %ld\n", rv);
+    blockq_handle_completion(ut->info.timerfd.bq, flags, bound(completion), t, rv);
+    closure_finish();
+    return rv;
+}
+
+closure_function(1, 6, sysreturn, timerfd_read,
+                 unix_timer, ut,
+                 void *, dest, u64, length, u64, offset_arg, thread, t, boolean, bh, io_completion, completion)
+{
+    if (length == 0)
+        return 0;
+    unix_timer ut = bound(ut);
+    timer_debug("fd %d, dest %p, length %ld, tid %d, bh %d, completion %p\n", ut->info.timerfd.fd,
+                dest, length, t->tid, bh, completion);
+    blockq_action ba = closure(unix_timer_heap, timerfd_read_bh, ut, dest, length, t, completion);
+    return blockq_check(ut->info.timerfd.bq, t, ba, bh);
+}
+
+closure_function(1, 1, u32, timerfd_events,
+                 unix_timer, ut,
+                 thread, t /* ignored */)
+{
+    return bound(ut)->overruns > 0 ? EPOLLIN : 0;
+}
+
+closure_function(1, 0, sysreturn, timerfd_close,
+                 unix_timer, ut)
+{
+    unix_timer ut = bound(ut);
+    remove_unix_timer(ut);
+    deallocate_blockq(ut->info.timerfd.bq);
+    deallocate_closure(ut->f.read);
```

Comment:
this is an opportunity for struct closure?

not really a comment, just curious

Reply:
Yes, as they are one-to-one. Since this is a pattern throughout our unix world, it would be best done in another PR.